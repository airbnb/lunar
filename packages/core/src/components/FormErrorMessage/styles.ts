import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ unit }) => ({
  error: {
    marginTop: unit,
  },
});

export default styleSheet;

export { styleSheet };
