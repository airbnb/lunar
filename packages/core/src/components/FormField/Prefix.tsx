import React from 'react';
import FieldAffix, { FieldAffixProps } from '../private/FieldAffix';

/** A prefix to display before an input within a form field. */
export default function Prefix({ children, small, large, disabled }: FieldAffixProps) {
  return (
    <FieldAffix before small={small} large={large} disabled={disabled}>
      {children}
    </FieldAffix>
  );
}
