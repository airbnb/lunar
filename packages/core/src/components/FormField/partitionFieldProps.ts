import { Props as FormFieldProps } from '.';

export type MaybeChildren = { children?: unknown };

export type ExplicitProps = {
  compact: boolean;
  disabled: boolean;
  hasPrefix: boolean;
  hasSuffix: boolean;
  invalid: boolean;
  large: boolean;
  optional: boolean;
  small: boolean;
  value: string;
};

export default function partitionFieldProps<Props extends MaybeChildren = {}>(
  props: FormFieldProps & Props,
): {
  // Need any for consumers to work correctly
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  children: any;
  field: object;
  fieldProps: FormFieldProps;
  inputProps: Omit<Props, 'children'> & ExplicitProps;
} {
  const {
    children,
    compact = false,
    compactSpacing = false,
    disabled = false,
    errorMessage = '',
    field = {},
    hideLabel = false,
    hideOptionalLabel = false,
    inline = false,
    invalid = false,
    label,
    labelDescription = '',
    large = false,
    noSpacing = false,
    optional = false,
    prefix = null,
    small = false,
    suffix = null,
    ...inputProps
  } = props;

  return {
    children,
    field,
    fieldProps: {
      compact,
      compactSpacing,
      disabled,
      errorMessage,
      hideLabel,
      hideOptionalLabel,
      inline,
      invalid,
      label,
      labelDescription,
      large,
      noSpacing,
      optional,
      prefix,
      small,
      suffix,
    },
    // @ts-ignore Hard to type
    inputProps: {
      value: '',
      ...inputProps,
      compact,
      disabled,
      hasPrefix: !!prefix,
      hasSuffix: !!suffix,
      invalid,
      large,
      optional,
      small,
    },
  };
}
