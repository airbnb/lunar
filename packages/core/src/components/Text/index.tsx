import React from 'react';
import { mutuallyExclusiveTrueProps } from 'airbnb-prop-types';
import withStyles, { css, WithStylesProps } from '../../composers/withStyles';

const sizingProp = mutuallyExclusiveTrueProps('micro', 'small', 'large');
const emphasisProp = mutuallyExclusiveTrueProps('bold', 'light');
const stateProp = mutuallyExclusiveTrueProps('disabled', 'muted', 'inverted');
const alignProp = mutuallyExclusiveTrueProps('centerAlign', 'endAlign', 'startAlign');

export type Props = {
  /** Render the text inline instead of block. */
  baseline?: boolean;
  /** Apply bold emphasis. */
  bold?: boolean;
  /** Align the text to the center. */
  centerAlign?: boolean;
  /** The text to render. */
  children?: React.ReactNode;
  /** Mark the text as disabled. */
  disabled?: boolean;
  /** Align the text to the end. */
  endAlign?: boolean;
  /** Render the text inline-block instead of block. */
  inline?: boolean;
  /** Invert text colors. */
  inverted?: boolean;
  /** Increase font size to large. */
  large?: boolean;
  /** Apply light emphasis. */
  light?: boolean;
  /** Decrease font size to small. */
  micro?: boolean;
  /** Mark the text as muted. */
  muted?: boolean;
  /** Disable text and white space wrapping. */
  noWrap?: boolean;
  /** Preserve whitespace at the beginning and end of children. */
  preserveWhitespace?: boolean;
  /** Decrease font size to small. */
  small?: boolean;
  /** Align the text to the start. */
  startAlign?: boolean;
  /** Truncate the text with an ellipsis. */
  truncated?: boolean;
  /** Uppercase all text. */
  uppercased?: boolean;
};

/** Display a string of text with pre-defined sizing, emphasis, and state styling. */
export class Text extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    bold: emphasisProp,
    centerAlign: alignProp,
    disabled: stateProp,
    endAlign: alignProp,
    inverted: stateProp,
    large: sizingProp,
    light: emphasisProp,
    micro: sizingProp,
    muted: stateProp,
    small: sizingProp,
    startAlign: alignProp,
  };

  static defaultProps = {
    baseline: false,
    bold: false,
    centerAlign: false,
    children: null,
    disabled: false,
    endAlign: false,
    inline: false,
    inverted: false,
    large: false,
    light: false,
    micro: false,
    muted: false,
    noWrap: false,
    preserveWhitespace: false,
    small: false,
    startAlign: false,
    truncated: false,
    uppercased: false,
  };

  render() {
    const {
      baseline,
      bold,
      centerAlign,
      children,
      disabled,
      endAlign,
      inline,
      inverted,
      large,
      light,
      micro,
      muted,
      noWrap,
      preserveWhitespace,
      small,
      startAlign,
      styles,
      truncated,
      uppercased,
    } = this.props;
    let Tag: 'div' | 'span' | 'small' | 'h4' = 'div';

    if (inline) {
      Tag = 'span';
    } else if (micro) {
      Tag = 'small';
    } else if (large) {
      Tag = 'h4';
    }

    return (
      <Tag
        {...css(
          styles.text,
          bold && styles.text_bold,
          disabled && styles.text_disabled,
          inline && styles.text_inline,
          baseline && styles.text_baseline,
          inverted && styles.text_inverted,
          large && styles.text_large,
          light && styles.text_light,
          micro && styles.text_micro,
          muted && styles.text_muted,
          preserveWhitespace && styles.text_preserveWhitespace,
          small && styles.text_small,
          truncated && styles.text_truncated,
          uppercased && styles.text_uppercased,
          centerAlign && styles.text_center,
          endAlign && styles.text_end,
          startAlign && styles.text_start,
          noWrap && styles.text_noWrap,
        )}
      >
        {children}
      </Tag>
    );
  }
}

export default withStyles(({ color, font, pattern }) => ({
  text: {
    ...font.textReset,
    ...font.textRegular,
    color: color.accent.text,
  },

  text_baseline: {
    display: 'inline',
  },

  text_inline: {
    display: 'inline-block',
  },

  text_preserveWhitespace: {
    whiteSpace: 'pre-wrap',
  },

  text_micro: {
    ...font.textMicro,
  },

  text_small: {
    ...font.textSmall,
  },

  text_large: {
    ...font.textLarge,
  },

  text_disabled: {
    ...pattern.disabled,
  },

  text_inverted: {
    color: color.base,
  },

  text_muted: {
    color: color.muted,
  },

  text_bold: {
    fontWeight: font.weights.semibold,
  },

  text_light: {
    fontWeight: font.weights.light,
  },

  text_truncated: {
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',

    '@selectors': {
      '> div': {
        display: 'inline',
      },
    },
  },

  text_uppercased: {
    textTransform: 'uppercase',
  },

  text_center: {
    textAlign: 'center',
  },

  text_end: {
    textAlign: 'right',
  },

  text_start: {
    textAlign: 'left',
  },

  text_noWrap: {
    whiteSpace: 'nowrap',
  },
}))(Text);
