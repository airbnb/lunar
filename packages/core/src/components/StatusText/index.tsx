import React from 'react';
import { mutuallyExclusiveTrueProps } from 'airbnb-prop-types';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import Text, { Props as TextProps } from '../Text';
import { STATUSES } from '../../constants';

const statusPropType = mutuallyExclusiveTrueProps(...STATUSES);

export type Props = TextProps & {
  /** The text to render. */
  children: NonNullable<React.ReactNode>;
  /** Dangerous/failure status (red). */
  danger?: boolean;
  /** Informational status (blue). */
  info?: boolean;
  /** Muted/disabled status (gray). */
  muted?: boolean;
  /** Notice status. */
  notice?: boolean;
  /** Successful status (green). */
  success?: boolean;
  /** Warning status (yellow). */
  warning?: boolean;
};

/** Display a string of classified text with colorful statuses. */
export class StatusText extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    danger: statusPropType,
    info: statusPropType,
    muted: statusPropType,
    notice: statusPropType,
    success: statusPropType,
    warning: statusPropType,
  };

  static defaultProps = {
    danger: false,
    info: false,
    muted: false,
    notice: false,
    success: false,
    warning: false,
  };

  render() {
    const {
      cx,
      children,
      danger,
      info,
      muted,
      notice,
      styles,
      success,
      warning,
      ...restProps
    } = this.props;

    return (
      <Text {...restProps}>
        <span
          className={cx(
            danger && styles.text_danger,
            info && styles.text_info,
            muted && styles.text_muted,
            notice && styles.text_notice,
            success && styles.text_success,
            warning && styles.text_warning,
          )}
        >
          {children}
        </span>
      </Text>
    );
  }
}

export default withStyles(({ color }) => ({
  text_danger: {
    color: color.core.danger[4],
  },

  text_info: {
    color: color.core.primary[4],
  },

  text_muted: {
    color: color.muted,
  },

  text_notice: {
    color: color.core.secondary[4],
  },

  text_success: {
    color: color.core.success[4],
  },

  text_warning: {
    color: color.core.warning[5],
  },
}))(StatusText);
