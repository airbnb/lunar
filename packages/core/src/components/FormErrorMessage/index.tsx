import React from 'react';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import StatusText from '../StatusText';

export type Props = {
  /** ID of the input field associated with the error. */
  id: string;
  /** The error message. */
  error?: string;
};

/** Display a simple error message for use within forms. */
export class FormErrorMessage extends React.Component<Props & WithStylesProps> {
  render() {
    const { cx, id, error, styles } = this.props;

    if (!error) {
      return null;
    }

    return (
      <div id={`${id}-error`} role="alert" aria-live="polite" className={cx(styles.error)}>
        <StatusText danger>{error}</StatusText>
      </div>
    );
  }
}

export default withStyles(({ unit }) => ({
  error: {
    marginTop: unit,
  },
}))(FormErrorMessage);
