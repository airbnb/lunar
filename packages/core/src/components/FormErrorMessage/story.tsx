import React from 'react';
import FormErrorMessage from '.';

export default {
  title: 'Core/FormErrorMessage',
  parameters: {
    inspectComponents: [FormErrorMessage],
  },
};

export function anErrorMessageForAnInvalidFormField() {
  return <FormErrorMessage id="foo" error="This field is required!" />;
}

anErrorMessageForAnInvalidFormField.story = {
  name: 'An error message for an invalid form field.',
};
