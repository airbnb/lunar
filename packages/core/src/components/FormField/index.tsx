import React from 'react';
import { childrenOfType } from 'airbnb-prop-types';
import withStyles, { css, WithStylesProps } from '../../composers/withStyles';
import T from '../Translate';
import Text from '../Text';
import StatusText from '../StatusText';
import FormErrorMessage from '../FormErrorMessage';
import partitionFieldProps from './partitionFieldProps';
import Prefix from './Prefix';
import Suffix from './Suffix';

export type Props = {
  /** Decrease label font size and spacing. */
  compact?: boolean;
  /** @ignore Decrease bottom margin of the field. (Internal use only) */
  compactSpacing?: boolean;
  /** Mark the field as disabled. */
  disabled?: boolean;
  /** Error message to display under the input when invalid. */
  errorMessage?: string;
  /** @ignore Passed from final form. */
  field?: any;
  /** Visually hide the label. */
  hideLabel?: boolean;
  /** @ignore Hide optional label. (Internal use only) */
  hideOptionalLabel?: boolean;
  /** Render the input and label inline to each other. */
  inline?: boolean;
  /** Mark the field as invalid. */
  invalid?: boolean;
  /** Descriptive label that appears above the input. */
  label: NonNullable<React.ReactNode>;
  /** Small description to display under the label. */
  labelDescription?: React.ReactNode;
  /** Remove bottom margin from field. */
  noSpacing?: boolean;
  /** Mark the field as optional. */
  optional?: boolean;
  /** Content to display before the input field. */
  prefix?: React.ReactNode;
  /** Content to display after the input field. */
  suffix?: React.ReactNode;
};

export type PrivateProps = Props &
  WithStylesProps & {
    /** Input field to wrap. */
    children: NonNullable<React.ReactNode>;
    /** Unique ID of the field. */
    id: string;
    /** @ignore Render the input field before the label. */
    renderBeforeLabel?: boolean;
    /** @ignore Render the input field as 100% width. */
    renderFullWidth?: boolean;
    /** @ignore Render the labels with large text. Should only be used for nested components, like through controllers. */
    renderLargeLabel?: boolean;
    /** @ignore Stretches the label to 100%. */
    stretchLabel?: boolean;
    /** @ignore Top align content. */
    topAlign?: boolean;
  };

export { partitionFieldProps, Prefix, Suffix };

/** A abstract form field wrapper that handles labels, affixes, errors, states, and more. */
export class FormField extends React.Component<PrivateProps> {
  static propTypes = {
    prefix: childrenOfType(Prefix),
    suffix: childrenOfType(Suffix),
  };

  static defaultProps = {
    compact: false,
    compactSpacing: false,
    disabled: false,
    errorMessage: '',
    hideLabel: false,
    hideOptionalLabel: false,
    inline: false,
    invalid: false,
    labelDescription: null,
    noSpacing: false,
    optional: false,
    prefix: null,
    renderBeforeLabel: false,
    renderFullWidth: false,
    renderLargeLabel: false,
    stretchLabel: false,
    suffix: null,
    topAlign: false,
  };

  render() {
    const {
      children,
      compact,
      compactSpacing,
      disabled,
      errorMessage,
      hideLabel,
      hideOptionalLabel,
      id,
      inline,
      invalid,
      label,
      labelDescription,
      noSpacing,
      optional,
      renderBeforeLabel,
      renderFullWidth,
      renderLargeLabel,
      stretchLabel,
      styles,
      prefix,
      suffix,
      topAlign,
    } = this.props;

    const content = (
      <div
        {...css(
          styles.input,
          inline && renderBeforeLabel && styles.input_beforeInline,
          inline && !renderBeforeLabel && styles.input_afterInline,
          hideLabel && styles.input_hideLabel,
          renderFullWidth && styles.input_fullWidth,
        )}
      >
        {prefix && <div {...css(styles.affix)}>{prefix}</div>}

        <div {...css(styles.anchor)}>{children}</div>

        {suffix && <div {...css(styles.affix)}>{suffix}</div>}
      </div>
    );

    return (
      <section
        {...css(
          styles.field,
          (compact || compactSpacing) && !noSpacing && styles.field_compactSpacing,
          noSpacing && styles.field_noSpacing,
        )}
      >
        <div {...css(inline && styles.content_inline, topAlign && styles.content_topAlign)}>
          {renderBeforeLabel && content}

          <label
            htmlFor={id}
            {...css(
              styles.label,
              stretchLabel && styles.label_stretch,
              hideLabel && styles.label_hidden,
              (inline || renderBeforeLabel) && styles.label_noSpacing,
            )}
          >
            <StatusText danger={invalid} muted={disabled} small={compact} bold={!renderLargeLabel}>
              {label}

              {optional && !hideOptionalLabel && (
                <span {...css(styles.optional)}>
                  <Text inline small muted>
                    <T phrase="(optional)" context="A form field is marked as optional" />
                  </Text>
                </span>
              )}
            </StatusText>

            {labelDescription && <Text small>{labelDescription}</Text>}
          </label>

          {!renderBeforeLabel && content}
        </div>

        {invalid && <FormErrorMessage id={id} error={errorMessage} />}
      </section>
    );
  }
}

export default withStyles(({ unit }) => ({
  field: {
    marginBottom: unit * 3,

    '@selectors': {
      ':last-child, :only-child': {
        marginBottom: 0,
      },
    },
  },

  field_compactSpacing: {
    marginBottom: unit * 2,
  },

  field_noSpacing: {
    margin: 0,
  },

  content_inline: {
    display: 'flex',
    alignItems: 'center',
  },

  content_topAlign: {
    alignItems: 'flext-start',
  },

  label: {
    margin: 0,
    marginBottom: unit,
    display: 'block',
    flexGrow: 0,
  },

  label_stretch: {
    flexBasis: '100%',
  },

  label_hidden: {
    display: 'none',
  },

  label_noSpacing: {
    margin: 0,
  },

  optional: {
    marginLeft: unit,
    display: 'inline-block',
  },

  input: {
    display: 'flex',
    position: 'relative',
    flexGrow: 1,
  },

  input_fullWidth: {
    width: '100%',
  },

  input_beforeInline: {
    marginRight: unit,
  },

  input_afterInline: {
    marginLeft: unit,
  },

  input_hideLabel: {
    margin: 0,
  },

  affix: {
    flexGrow: 0,
  },

  anchor: {
    flexGrow: 1,
  },
}))(FormField);
