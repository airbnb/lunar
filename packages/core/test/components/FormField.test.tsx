import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import T from '../../src/components/Translate';
import FormField, { Prefix, Suffix } from '../../src/components/FormField';
import FormErrorMessage from '../../src/components/FormErrorMessage';
import StatusText from '../../src/components/StatusText';
import Text from '../../src/components/Text';

describe('<FormField />', () => {
  it('renders a label', () => {
    const wrapper = shallowWithStyles(
      <FormField id="foo" label="Label">
        Foo
      </FormField>,
    );

    expect(wrapper.find('label')).toHaveLength(1);
    expect(wrapper.find(StatusText).prop('children')).toEqual(['Label', false]);
  });

  it('renders a label with optional copy', () => {
    const wrapper = shallowWithStyles(
      <FormField id="foo" label="Label">
        Foo
      </FormField>,
    );

    expect(wrapper.find('label').find(T)).toHaveLength(0);

    wrapper.setProps({
      optional: true,
    });

    expect(wrapper.find('label').find(T)).toHaveLength(1);
    expect(
      wrapper
        .find('label')
        .find(T)
        .prop('phrase'),
    ).toBe('(optional)');
  });

  it('doesnt renders optional label if `hideOptionalLabel` is true', () => {
    const wrapper = shallowWithStyles(
      <FormField id="foo" label="Label" hideOptionalLabel>
        Foo
      </FormField>,
    );

    expect(wrapper.find('label').find(T)).toHaveLength(0);

    wrapper.setProps({
      optional: true,
    });

    expect(wrapper.find('label').find(T)).toHaveLength(0);
  });

  it('renders a label description', () => {
    const wrapper = shallowWithStyles(
      <FormField id="foo" label="Label">
        Foo
      </FormField>,
    );

    expect(wrapper.find(Text)).toHaveLength(0);

    wrapper.setProps({
      labelDescription: 'Description',
    });

    expect(wrapper.find(Text)).toHaveLength(1);
    expect(wrapper.find(Text).prop('children')).toBe('Description');
  });

  it('renders input after the label', () => {
    const wrapper = shallowWithStyles(
      <FormField id="foo" label="Label">
        Foo
      </FormField>,
    );

    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({
      inline: true,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('renders input before the label', () => {
    const wrapper = shallowWithStyles(
      <FormField id="foo" label="Label" renderBeforeLabel>
        Foo
      </FormField>,
    );

    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({
      inline: true,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('hides the label', () => {
    const wrapper = shallowWithStyles(
      <FormField id="foo" label="Label" hideLabel>
        Foo
      </FormField>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('shows error if invalid', () => {
    const wrapper = shallowWithStyles(
      <FormField id="foo" label="Label" invalid errorMessage="Broken!">
        Foo
      </FormField>,
    );

    expect(wrapper.find(FormErrorMessage).prop('error')).toBe('Broken!');

    wrapper.setProps({
      invalid: false,
    });

    expect(wrapper.find(FormErrorMessage)).toHaveLength(0);
  });

  it('supports a prefix', () => {
    const wrapper = shallowWithStyles(
      <FormField id="foo" label="Label" prefix={<Prefix>Foo</Prefix>}>
        Foo
      </FormField>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('errors for invalid prefix', () => {
    expect(() => {
      shallowWithStyles(
        <FormField id="foo" label="Label" prefix={<div />}>
          Foo
        </FormField>,
      );
    }).toThrowErrorMatchingSnapshot();
  });

  it('supports a suffix', () => {
    const wrapper = shallowWithStyles(
      <FormField id="foo" label="Label" suffix={<Suffix>Foo</Suffix>}>
        Foo
      </FormField>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('errors for invalid suffix', () => {
    expect(() => {
      shallowWithStyles(
        <FormField id="foo" label="Label" suffix={<div />}>
          Foo
        </FormField>,
      );
    }).toThrowErrorMatchingSnapshot();
  });
});
