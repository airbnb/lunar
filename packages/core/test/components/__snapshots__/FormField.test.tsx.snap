// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormField /> errors for invalid prefix 1`] = `
"Warning: Failed prop type: \`FormField\` only accepts children of type Prefix
    in FormField"
`;

exports[`<FormField /> errors for invalid suffix 1`] = `
"Warning: Failed prop type: \`FormField\` only accepts children of type Suffix
    in FormField"
`;

exports[`<FormField /> hides the label 1`] = `
<section
  className="field_k3wq0g"
>
  <div
    className=""
  >
    <label
      className="label_2s5uhd-o_O-label_hidden_1u9fru1"
      htmlFor="foo"
    >
      <StatusText
        bold={true}
        danger={false}
        muted={false}
        small={false}
      >
        Label
      </StatusText>
    </label>
    <div
      className="input_nv7oj6-o_O-input_hideLabel_16pg94n"
    >
      <div
        className="anchor_t0tx82"
      >
        Foo
      </div>
    </div>
  </div>
</section>
`;

exports[`<FormField /> renders input after the label 1`] = `
<section
  className="field_k3wq0g"
>
  <div
    className=""
  >
    <label
      className="label_2s5uhd"
      htmlFor="foo"
    >
      <StatusText
        bold={true}
        danger={false}
        muted={false}
        small={false}
      >
        Label
      </StatusText>
    </label>
    <div
      className="input_nv7oj6"
    >
      <div
        className="anchor_t0tx82"
      >
        Foo
      </div>
    </div>
  </div>
</section>
`;

exports[`<FormField /> renders input after the label 2`] = `
<section
  className="field_k3wq0g"
>
  <div
    className="content_inline_5kaapu"
  >
    <label
      className="label_2s5uhd-o_O-label_noSpacing_16pg94n"
      htmlFor="foo"
    >
      <StatusText
        bold={true}
        danger={false}
        muted={false}
        small={false}
      >
        Label
      </StatusText>
    </label>
    <div
      className="input_nv7oj6-o_O-input_afterInline_pog3hg"
    >
      <div
        className="anchor_t0tx82"
      >
        Foo
      </div>
    </div>
  </div>
</section>
`;

exports[`<FormField /> renders input before the label 1`] = `
<section
  className="field_k3wq0g"
>
  <div
    className=""
  >
    <div
      className="input_nv7oj6"
    >
      <div
        className="anchor_t0tx82"
      >
        Foo
      </div>
    </div>
    <label
      className="label_2s5uhd-o_O-label_noSpacing_16pg94n"
      htmlFor="foo"
    >
      <StatusText
        bold={true}
        danger={false}
        muted={false}
        small={false}
      >
        Label
      </StatusText>
    </label>
  </div>
</section>
`;

exports[`<FormField /> renders input before the label 2`] = `
<section
  className="field_k3wq0g"
>
  <div
    className="content_inline_5kaapu"
  >
    <div
      className="input_nv7oj6-o_O-input_beforeInline_14tkmhr"
    >
      <div
        className="anchor_t0tx82"
      >
        Foo
      </div>
    </div>
    <label
      className="label_2s5uhd-o_O-label_noSpacing_16pg94n"
      htmlFor="foo"
    >
      <StatusText
        bold={true}
        danger={false}
        muted={false}
        small={false}
      >
        Label
      </StatusText>
    </label>
  </div>
</section>
`;

exports[`<FormField /> supports a prefix 1`] = `
<section
  className="field_k3wq0g"
>
  <div
    className=""
  >
    <label
      className="label_2s5uhd"
      htmlFor="foo"
    >
      <StatusText
        bold={true}
        danger={false}
        muted={false}
        small={false}
      >
        Label
      </StatusText>
    </label>
    <div
      className="input_nv7oj6"
    >
      <div
        className="affix_1n0jrn7"
      >
        <Prefix>
          Foo
        </Prefix>
      </div>
      <div
        className="anchor_t0tx82"
      >
        Foo
      </div>
    </div>
  </div>
</section>
`;

exports[`<FormField /> supports a suffix 1`] = `
<section
  className="field_k3wq0g"
>
  <div
    className=""
  >
    <label
      className="label_2s5uhd"
      htmlFor="foo"
    >
      <StatusText
        bold={true}
        danger={false}
        muted={false}
        small={false}
      >
        Label
      </StatusText>
    </label>
    <div
      className="input_nv7oj6"
    >
      <div
        className="anchor_t0tx82"
      >
        Foo
      </div>
      <div
        className="affix_1n0jrn7"
      >
        <Suffix>
          Foo
        </Suffix>
      </div>
    </div>
  </div>
</section>
`;
